src/index/src/inverted_index/create/sort/external_sort.rs

```diff
@@ -0,0 +1,383 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::btree_map::Entry;
+use std::collections::BTreeMap;
+use std::mem;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use common_base::BitVec;
+use futures::stream;
+
+use crate::inverted_index::create::sort::external_provider::ExternalTempFileProvider;
+use crate::inverted_index::create::sort::intermediate_rw::{
+    IntermediateReader, IntermediateWriter,
+};
+use crate::inverted_index::create::sort::merge_stream::MergeSortedStream;
+use crate::inverted_index::create::sort::{SortOutput, SortedStream, Sorter};
+use crate::inverted_index::error::Result;
+use crate::inverted_index::Bytes;
+
+/// `ExternalSorter` manages the sorting of data using both in-memory structures and external files.
+/// It dumps data to external files when the in-memory buffer crosses a certain memory threshold.
+pub struct ExternalSorter {
+    /// The index name associated with the sorting operation
+    index_name: String,
+
+    /// Manages creation and access to external temporary files
+    temp_file_provider: Arc<dyn ExternalTempFileProvider>,
+
+    /// Bitmap indicating which segments have null values
+    null_bitmap: BitVec,
+
+    /// In-memory buffer to hold values and their corresponding bitmaps until memory threshold is exceeded
+    values_buffer: BTreeMap<Bytes, BitVec>,
+
+    /// Count of all rows ingested so far
+    total_row_count: usize,
+
+    /// The number of rows per group for bitmap indexing which determines how rows are
+    /// batched for indexing. It is used to determine which segment a row belongs to.
+    segment_row_count: usize,
+
+    /// Tracks memory usage of the buffer
+    current_memory_usage: usize,
+
+    /// The memory usage threshold at which the buffer should be dumped to an external file
+    memory_usage_threshold: usize,
```

Comment:
Do we need to define the threshold as `Option<usize>` and use `None` to disable external sort? If we want to disable this by default, the TOML rendered might contain a very large number.